cassandra-executor/src/main/java/com/mesosphere/dcos/cassandra/executor/compress/SnappyCompressedChunker.java

```diff
@@ -0,0 +1,111 @@
+package com.mesosphere.dcos.cassandra.executor.compress;
+
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xerial.snappy.SnappyOutputStream;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Iterator;
+
+/**
+ * SnappyCompressedChunker reads bytes from InputStream, compress them, and return a stream of byte[]. The size of
+ * byte[] is equal to compressedChunkSize, for all but the last chunk.
+ */
+public class SnappyCompressedChunker implements Iterator<byte[]> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnappyCompressedChunker.class);
+
+    /**
+     * Number of bytes to read from disk.
+     */
+    private static int READ_BUFFER_SIZE = 4 * 1024 * 1024; // 4MB
+
+    /**
+     * Amount of compressed bytes to buffer.
+     */
+    private int compressedChunkSize;
+
+    private boolean hasNextChunk;
+    private ByteArrayOutputStream baos;
+    private InputStream fileStreamToCompress;
+    private SnappyOutputStream compressedStream;
+
+    public SnappyCompressedChunker(final InputStream fileToCompress,
+                                   final int compressedChunkSize) throws IOException {
+        this.fileStreamToCompress = fileToCompress;
+        this.compressedChunkSize = compressedChunkSize;
+        this.hasNextChunk = true;
+
+        // Initialize Snappy compression stream
+        this.baos = new ByteArrayOutputStream();
+        this.compressedStream = new SnappyOutputStream(baos);
+    }
+
+    @Override
+    public boolean hasNext() {
+        return hasNextChunk;
+    }
+
+    @Override
+    public byte[] next() {
```

Comment:
I don't think this is what you want to do here. This is reading 4 MB of uncompressed data and then writing it to a buffer that will contain the compressed data. Unless the entropy of the request is 0, the returned buffer will not be a 4 MB chunk.